---
title: "Obtain Validation input data required for cell-cell communication random graphs"
output: html_document
vignette: >
  %\VignetteIndexEntry{easier User Manual}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

<style type="text/css">
.smaller {
  font-size: 10px
  font-style: normal
}
</style>

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, include = FALSE}
# working directory
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

# packages
library('corrplot')
library('readxl')  
library('reshape2')
library('dplyr')

# Change default settings to fit plots
knitr::opts_chunk$set(fig.height = 9, fig.width = 9)

## path to validation data folder
folderValidation <- "/Users/Oscar/ownCloud2/SystemsImmunoOncology/Mechanistic_signatures_project/data/Validation/prediction_files"

## path to curated cell-cell communication table
path_to_CCCtable <- "../Data/CCC_PKN/expressed_ccc_table.csv"

## path to output folder
path_to_output <- ""

```

> Cell-cell communication table obtained from `RaCInG_ccc_prior_knowledge.Rmd`

```{r}

# curated cell-cell communication table
ccc_table_filtered <- read.csv(file = path_to_CCCtable)

# exract unique ligand-receptor pairs
LR_pairs <- unique(paste0(ccc_table_filtered$Ligand, "_", ccc_table_filtered$Receptor))

```

> Compute ligand-receptor pair weights

```{r}

datasets_bulk <-c("gideauslanderpd1" = "SKCM",
                  "gideauslanderpd1on" = "SKCM",
                  "kim" =  "STAD")

for (dataset in names(datasets_bulk)){
  
  cat("Dataset:", dataset, "\n")
  # get cancer type
  cancer.type <- as.character(datasets_bulk[dataset])
  
  # load folder
  load(paste0(folderValidation, "/", dataset, "/all_", dataset, ".RData"))
    
  # Gene expression data
  RNA_tpm <- get(paste0("folder_", dataset))$tpm
  
  # log2 transform
  log2_RNA_tpm <- as.data.frame(log2(RNA_tpm + 1))
  
  ## using limiting factor strategy: min(L,R) --> compute L-R pairs weights
  LR_pairs_computed <- do.call(rbind, lapply(
    seq_len(length(LR_pairs)),
    function(x) {
      ligand <- vapply(strsplit(LR_pairs[x], split = "_", fixed = TRUE), head, 1,
                       FUN.VALUE = character(1)
      )
      receptor <- vapply(strsplit(LR_pairs[x], split = "_", fixed = TRUE), tail, 1,
                         FUN.VALUE = character(1)
      )
      pos_lr <- match(c(ligand, receptor), rownames(log2_RNA_tpm))
      # When a ligand or receptor is not found, NA value should be returned.
      by_patient <- t(as.data.frame(apply(log2_RNA_tpm[pos_lr, ], 2, min)))
      rownames(by_patient) <- LR_pairs[x]
      return(by_patient)
    }
  ))
  colnames(LR_pairs_computed) <- gsub(".", "-", colnames(LR_pairs_computed), fixed = TRUE)
  LR_pairs_computed <- t(LR_pairs_computed)
  
  write.csv(LR_pairs_computed, file = file.path(path_to_output, dataset, "LRpair_weights_min.csv"))
  
}

```

> Compute cell-type abundance

```{r}

# functions to compute consensus cell fractions
source("./utils/run_TMEmod_deconvolution.R")
range01 <- function(x){(x-min(x))/(max(x)-min(x))}

datasets_bulk <-c("gideauslanderpd1" = "SKCM",
                  "gideauslanderpd1on" = "SKCM",
                  "kim" =  "STAD")

for (dataset in names(datasets_bulk)){
  
  cat("Dataset:", dataset, "\n")
  # get cancer type
  cancer.type <- as.character(datasets_bulk[dataset])
  
  # load folder
  load(paste0(folderValidation, "/all_", dataset, ".RData"))
    
  # Gene expression data
  RNA_tpm <- get(paste0("folder_", dataset))$tpm
  
  # Compute cell fractions
  cellfrac <- run_TMEmod_deconvolution(RNA_tpm, cancer_type = cancer.type)
  
  cellfrac.quanTIseq <- cellfrac$quanTIseq
  cellfrac.TIMER <- cellfrac$TIMER
  cellfrac.consensusTME <- cellfrac$consensusTME
  cellfrac.xCell <-cellfrac$xCell
  cellfrac.mcpcounter <- cellfrac$mcpcounter
  cellfrac.EPIC <- cellfrac$EPIC
  
  ## Load deconvolution results
  DCres <- data.frame(CD8_qTs = cellfrac.quanTIseq["T.cells.CD8",],
                      NK_qTs = cellfrac.quanTIseq["NK.cells",],
                      DC_qTs = cellfrac.quanTIseq["Dendritic.cells",],
                      Treg_qTs = cellfrac.quanTIseq["Tregs",],
                      Tumor_qTs = cellfrac.quanTIseq["Other",],
                      B_qTs = cellfrac.quanTIseq["B.cells",],
                      M1_qTs = cellfrac.quanTIseq["Macrophages.M1",],
                      M2_qTs = cellfrac.quanTIseq["Macrophages.M2",],
                      CD8_TIM = cellfrac.TIMER["T_cell.CD8",],
                      DC_TIM = cellfrac.TIMER["DC",],
                      B_TIM = cellfrac.TIMER["B_cell",],
                      M1_con = cellfrac.consensusTME["Macrophages_M1",],
                      M2_con = cellfrac.consensusTME["Macrophages_M2",],
                      CD8_con = cellfrac.consensusTME["T_cells_CD8",],
                      DC_con = cellfrac.consensusTME["Dendritic_cells",],
                      Endo_con = cellfrac.consensusTME["Endothelial",],
                      CAF_con = cellfrac.consensusTME["Fibroblasts",],
                      Treg_con = cellfrac.consensusTME["T_regulatory_cells",],
                      NK_con = cellfrac.consensusTME["NK_cells",],
                      B_con = cellfrac.consensusTME["B_cells",],
                      Endo_xCl = cellfrac.xCell["Endothelial cells",],
                      Adip_xCl = cellfrac.xCell["Adipocytes",],
                      B_xCl = cellfrac.xCell["B-cells",],
                      DC_xCl = cellfrac.xCell["DC",],
                      CD8_xCl = cellfrac.xCell["CD8+ T-cells",],
                      CAF_xCl = cellfrac.xCell["Fibroblasts",],
                      NK_xCl = cellfrac.xCell["NK cells",],
                      Treg_xCl = cellfrac.xCell["Tregs",],
                      M1_xCl = cellfrac.xCell["Macrophages M1",],
                      M2_xCl = cellfrac.xCell["Macrophages M2",],
                      B_MCP = cellfrac.mcpcounter["B lineage",],
                      CD8_MCP = cellfrac.mcpcounter["CD8 T cells",],
                      NK_MCP = cellfrac.mcpcounter["NK cells",],
                      DC_MCP = cellfrac.mcpcounter["Myeloid dendritic cells",],
                      CAF_MCP = cellfrac.mcpcounter["Fibroblasts",],
                      CD8_EPI = cellfrac.EPIC["CD8_Tcells",],
                      B_EPI = cellfrac.EPIC["Bcells",],
                      NK_EPI = cellfrac.EPIC["NKcells",],
                      CAF_EPI = cellfrac.EPIC["CAFs",],
                      Tumor_EPI = cellfrac.EPIC["otherCells",],
                      Endo_EPI = cellfrac.EPIC["Endothelial",])
  
  DCres <- DCres[, order(colnames(DCres))]
  DCres.cor <- round(cor(DCres),2)
  
  # Save results
  celltypes <- c("DC", "CD8", "Treg", 
                 "B", "M1", "M2",
                 "Endo", "CAF", "NK", "Tumor")
  CellFrac <- as.data.frame(matrix(NA, 
                                   ncol = length(celltypes),
                                   nrow = nrow(DCres)))
  rownames(CellFrac) <- rownames(DCres)
  colnames(CellFrac) <- celltypes
  
  ## CD8 T cells
  CD8_cor <-DCres.cor["CD8_qTs", -grep("_qTs$", colnames(DCres.cor))]
  (CD8_cor <- sort(CD8_cor[grep("^CD8_", names(CD8_cor))])) 
  # --> CD8_qTs has good agreement with all methods
  # --> Take CD8_qTs
  plot(DCres$CD8_qTs, DCres$CD8_EPI); abline(0,1) # offset
  CellFrac$CD8 <- DCres$CD8_qTs
  
  ## CAFs
  CAF_cor <- DCres.cor["CAF_EPI", -grep("_EPI$", colnames(DCres.cor))]
  (CAF_cor <- sort(CAF_cor[grep("^CAF_", names(CAF_cor))]) )
  # --> CAF_EPI has good agreement with all methods
  # --> Take CAF_EPI
  CellFrac$CAF <- DCres$CAF_EPI
  
  ## B cells
  B_cor <- DCres.cor["B_qTs", -grep("_qTs$", colnames(DCres.cor))]
  (B_cor <- sort(B_cor[grep("^B_", names(B_cor))]))
  # --> B_qTs has good agreement with all methods
  # --> Take B_qTs
  plot(DCres$B_qTs, DCres$B_EPI); abline(0,1)
  CellFrac$B <- DCres$B_qTs
  
  ## Natural Killer cells
  NK_cor <- DCres.cor["NK_qTs", -grep("_qTs$", colnames(DCres.cor))]
  (NK_cor <- sort(NK_cor[grep("^NK_", names(NK_cor))]))
  # --> quanTIseq has a low agreement with all methods
  NK_cor <- DCres.cor["NK_EPI", -grep("_EPI$", colnames(DCres.cor))]
  (NK_cor <- sort(NK_cor[grep("^NK_", names(NK_cor))]))
  # --> NK_EPI has a mid-low agreement with NK_MCP, NK_xCell, and NK_con
  # --> Take NK_EPI and set to 0 samples that are 0 for xCell
  CellFrac$NK <- DCres$NK_EPI
  CellFrac$NK[DCres$xCell == 0] <- 0
  
  ## M1 macrophages
  M1_cor <- DCres.cor["M1_qTs", -grep("_qTs$", colnames(DCres.cor))]
  (M1_cor <- sort(M1_cor[grep("^M1_", names(M1_cor))]))
  # --> M1_qTs has a mid agreement with M1_con and M1_xCl
  # --> Take M1_qTs
  CellFrac$M1 <- DCres$M1_qTs
  
  ## M2 macrophages
  M2_cor <- DCres.cor["M2_qTs", -grep("_qTs$", colnames(DCres.cor))]
  (M2_cor <- sort(M2_cor[grep("^M2_", names(M2_cor))]))
  # --> M2_qTs has a mid agreement with M2_con and M2_xCl
  # --> Take M2_qTs
  CellFrac$M2 <- DCres$M2_qTs
  
  ## Dendritic cells
  DC_cor <- DCres.cor["DC_qTs", -grep("_qTs$", colnames(DCres.cor))]
  (DC_cor <- sort(DC_cor[grep("^DC_", names(DC_cor))]))
  round(sum(DCres$DC_qTs == 0)/nrow(DCres)*100,2)
  # --> DC_qTs has a low agreement with all methods and many 0's (97.1%)
  DCres.cor["DC_TIM", grep("^DC_", colnames(DCres.cor))]
  # --> DC_TIM has good agreement with DC_MCP and DC_xCl
  # --> Use DC_TIM to predict quanTIseq scores for absent DC
  # DCres.DC <- DCres[DCres$DC_qTs !=0, ]
  # DC.lm <- lm(DC_qTs ~ DC_TIM, data = DCres.DC)
  # newdata <- data.frame(DC_TIM = DCres$DC_TIM)
  # CellFrac$DC <- predict(DC.lm, newdata = newdata)
  # plot(DCres$DC_TIM, DCres$DC_qTs)
  # plot(DCres$DC_TIM, CellFrac$DC)
  # --> Use a consensus of the most robust approaches
  # rescaled to the quanTIseq range (with xCell 0's)
  DCcon <- apply(cbind(range01(DCres$DC_xCl),
                       range01(DCres$DC_MCP),
                       range01(DCres$DC_TIM)),
                 1, median)
  DCrange <- DCres$DC_qTs
  DCrange[which(DCres$DC_xCl == 0)] <- 0
  DCrange <- range(DCrange)  
  CellFrac$DC <- DCcon * (DCrange[2] - DCrange[1]) + DCrange[1]
  plot(DCres$DC_MCP, CellFrac$DC)
  
  ## Regulatory T cells
  Treg_cor <- DCres.cor["Treg_qTs", -grep("_qTs$", colnames(DCres.cor))]
  (Treg_cor <- sort(Treg_cor[grep("^Treg_", names(Treg_cor))]))
  # --> Treg_qTs has a mid agreement with Treg_xCl and Treg_con
  # --> But usually overestimates Tregs
  # --> Take Treg_qTs but set to 0 samples that are 0 for xCell
  CellFrac$Treg <- DCres$Treg_qTs
  CellFrac$Treg[DCres$Treg_xCl == 0] <- 0
  plot(DCres$Treg_qTs, DCres$Treg_xCell)
  plot(DCres$Treg_qTs, DCres$Treg_con)
  plot(DCres$Treg_qTs, CellFrac$Treg)
  
  ## Endothelial cells
  Endo_cor <- DCres.cor["Endo_EPI", -grep("_EPI$", colnames(DCres.cor))]
  (Endo_cor <- sort(Endo_cor[grep("^Endo_", names(Endo_cor))]))
  # --> Endo_EPI has high agreement with Endo_con and Endo_xCl
  # --> Take Endo_EPI
  CellFrac$Endo <- DCres$Endo_EPI
  
  ## Tumor cells
  Tumor_cor <- DCres.cor["Tumor_qTs", -grep("_qTs$", colnames(DCres.cor))]
  (Tumor_cor <- sort(Tumor_cor[grep("^Tumor_", names(Tumor_cor))]))
  plot(DCres$Tumor_EPI, DCres$Tumor_qTs); abline(0,1)
  # --> Tumor_qTs and Tumor_EPI have high agreement
  # --> Take EPIC (no Endo, no CAFs)
  CellFrac$Tumor <- DCres$Tumor_EPI
  
  ## Normalize to sum to 100%
  CellFrac.norm <- CellFrac
  CellFrac.norm <- CellFrac.norm/apply(CellFrac.norm, 1, sum)
  
  ## Check correlation of the final score
  DCres.all <- cbind(DCres, CellFrac.norm)
  DCres.all <- DCres.all[, order(colnames(DCres.all))]
  DCres.cor <- cor(DCres.all)
  ucells <- unique(gsub("_.*$", "", rownames(DCres.cor)))
  for (cell in ucells) {
    
    i <- which(gsub("_.*$", "", rownames(DCres.cor)) == cell )
    if (length(i) == 1) next
    cat("\n")
    res <- DCres.cor[i,i]
    res <- res[-grep("_", rownames(res)),]
    res <- sort(res[grep("_", names(res))])
    print(res)
    
  }

  #corrplot(DCres.cor,
  #         tl.col = "black",
  #         tl.cex = 0.7)
  #order = "hclust",
  #hclust.method = "ward.D2"

  #boxplot(CellFrac.norm, las = 2)
  
  # to .csv file
  write.csv(CellFrac.norm,
            file= file.path(path_to_output, dataset, "TMEmod_cell_fractions.csv"), 
            row.names = TRUE)
}

```

> Obtain clean metadata for validation cohorts: immuntherapy response, immune subtyping from bagaev, OS, PFS

```{r}

## melanoma
MFP_melanoma <- read.csv("../../data/Bagaev_MFP_subtyping/annotation-panmi.tsv", header=TRUE, sep = "\t")
MFP_melanoma <- MFP_melanoma[MFP_melanoma$Cohort %in% c("Hugo", "Riaz", "Liu", "Auslander", "Gide"), ]
MFP_melanoma <- MFP_melanoma[,c("Sample", "Cohort", "MFP", "Patient", "Gender", "Age", "Diagnosis", 
                                "Therapy", "Prior_aCTLA4", "Prior_MAPKi","Checkpoint_Therapy_Design", 
                                "Time", "Days2CTLA4", "Days2PD1",
                                "Recist", "Response", "PFS", "PFS_FLAG", "OS", "OS_FLAG",
                                "Mutation_Subtype", "Mutation_load", "Mutation_Status",            
                                "Purity", "Storage_method")]

# ---------------------- #
## gideauslanderpd1 ##
# ---------------------- #
MFP_gideauslanderpd1 <- MFP_melanoma %>% filter(Cohort %in% c("Auslander", "Gide"))
load("../../../gene_panel/data/validation/bulk/all_validation_datasets/all_gideauslanderpd1.RData")
MFP_gideauslanderpd1 <- MFP_gideauslanderpd1[match(folder_gideauslanderpd1$clinical_response$Sample, MFP_gideauslanderpd1$Sample),]

MFP_gideauslanderpd1 <- MFP_gideauslanderpd1[,c("Sample", "MFP", "Recist", "Response", "OS", "OS_FLAG", "Prior_aCTLA4", "Prior_MAPKi", "Checkpoint_Therapy_Design")]
MFP_gideauslanderpd1$Prior_aCTLA4[MFP_gideauslanderpd1$Prior_aCTLA4 == ""] <- "No"
MFP_gideauslanderpd1$Prior_MAPKi[MFP_gideauslanderpd1$Prior_MAPKi == ""] <- "No"

load("../../data/random_graphs_input/validation/skcm_gideauslanderpd1/GideAuslander_TMEmod_cell_fractions.RData")
rownames(MFP_gideauslanderpd1) <- MFP_gideauslanderpd1$Sample
MFP_gideauslanderpd1 <- MFP_gideauslanderpd1[match(rownames(CellFrac.norm), MFP_gideauslanderpd1$Sample),]
MFP_gideauslanderpd1$Sample <- NULL

write.csv(MFP_gideauslanderpd1, file = "../../data/random_graphs_input/validation/validation_data_for_mike/metadata_gideauslanderpd1.csv", row.names = FALSE)

# ---------------------- #
## gideauslanderpd1on ##
# ---------------------- #
MFP_gideauslanderpd1on <- MFP_melanoma %>% filter(Cohort %in% c("Auslander", "Gide"))
load("../../../gene_panel/data/validation/bulk/all_validation_datasets/all_gideauslanderpd1_ontreatment.RData")
MFP_gideauslanderpd1on <- MFP_gideauslanderpd1on[match(folder_gideauslanderpd1_on$clinical_response$Sample, MFP_gideauslanderpd1on$Sample),]

MFP_gideauslanderpd1on <- MFP_gideauslanderpd1on[,c("Sample", "MFP", "Recist", "Response", "OS", "OS_FLAG", "Prior_aCTLA4", "Prior_MAPKi", "Checkpoint_Therapy_Design")]
MFP_gideauslanderpd1on$Prior_aCTLA4[MFP_gideauslanderpd1on$Prior_aCTLA4 == ""] <- "No"
MFP_gideauslanderpd1on$Prior_MAPKi[MFP_gideauslanderpd1on$Prior_MAPKi == ""] <- "No"

load("../../data/random_graphs_input/validation/skcm_gideauslanderpd1_on/gideauslanderpd1on_TMEmod_cell_fractions.RData")
rownames(MFP_gideauslanderpd1on) <- MFP_gideauslanderpd1on$Sample
MFP_gideauslanderpd1on <- MFP_gideauslanderpd1on[match(rownames(CellFrac.norm), MFP_gideauslanderpd1on$Sample),]
MFP_gideauslanderpd1on$Sample <- NULL
write.csv(MFP_gideauslanderpd1on, file = "../../data/random_graphs_input/validation/validation_data_for_mike/metadata_gideauslanderpd1on.csv", row.names = FALSE)

# ---------------------- #
## gastric cancer (kim) ##
# ---------------------- #
load("../../data/random_graphs_input/validation/stad_kim/kim_TMEmod_cell_fractions.RData")
load("~/ownCloud2/SystemsImmunoOncology/Mechanistic_signatures_project/data/Validation/Francesca/Kim/Kim_sampleinfo.rdata")
pre.sampleinfo <- subset(sampleinfo, Assay.Type == "RNA-Seq")
pre.sampleinfo <- pre.sampleinfo[match(rownames(CellFrac.norm), rownames(pre.sampleinfo)), ]
pre.sampleinfo$Alias <- substr(pre.sampleinfo$Alias, 1, 9)

MFP_gastric <- read.csv("../../data/Bagaev_MFP_subtyping/annotation-gastric.tsv", header=TRUE, sep = "\t")
MFP_gastric <- MFP_gastric[,c("Patient", "MFP", "Recist", "Response", "Mutation_load", "Pathology", "PDL_Tumor", "PDL_Immune", "PDL_CPS")]
MFP_gastric <- MFP_gastric[match(pre.sampleinfo$Alias, MFP_gastric$Patient),]
rownames(MFP_gastric) <- rownames(pre.sampleinfo)
  
MFP_gastric <- MFP_gastric[match(rownames(CellFrac.norm), rownames(MFP_gastric)),]
write.csv(MFP_gastric, file = "../../data/random_graphs_input/validation/validation_data_for_mike/metadata_kim.csv", row.names = FALSE)

```
